[core] Limit pitch based on edge insets. Fix max Z calculation in getProjMatrix.

astojilj opened the pull request:
Patch partly fixes #15163 in a way that it doesn't allow loading tens of thousands of tiles and attempt  to show area above horizon:
 
1. Limit pitch based on edge insets. It is not too bad - current limit of 60 degrees stays active until center of perspective is moved towards the bottom, to 84% of screen height. The plan is to split removal of 60 degrees limit to follow up patch.
2. Fix max Z calculation in getProjMatrix. TransformState::getProjMatrix calculation of farZ was complex with possibility to lead to negative z values. Replacing it with simpler, precise calculation:

``` furthestDistance = cameraToCenterDistance / (1 - tanFovAboveCenter * std::tan(getPitch())); ```

![IMG_0254](https://user-images.githubusercontent.com/549216/61720853-33e37d80-ad70-11e9-897c-bdb6c9a07f7f.JPG)

Mapping diagram to code:
fD: furthestDistance,
c: cameraToCenterDistance,
small phi: getPitch()




Addresses: #15163

1ec5 commented on src/mbgl/map/transform.cpp:
```diff
@@ -576,4 +578,14 @@ LatLng Transform::screenCoordinateToLatLng(const ScreenCoordinate& point, LatLng
     return state.screenCoordinateToLatLng(flippedPoint, wrapMode);
 }
 
+double Transform::getMaxPitchForEdgeInsets(const EdgeInsets &insets) const
```
How does this method relate to `PITCH_MAX`? Should we incorporate that constant somehow or remove it?

https://github.com/mapbox/mapbox-gl-native/blob/cafecd3f6fedc16e7bf5dff699960978c3b5cb3d/include/mbgl/util/constants.hpp#L37

astojilj replied:
I plan to remove it - which would mean that with no edge inset, getMaxPitchForEdgeInsets would evaluate to 71 degrees, if I remember correctly.

astojilj replied:
> I plan to remove it - which would mean that with no edge inset, getMaxPitchForEdgeInsets would evaluate to 71 degrees, if I remember correctly.

Edit: it is part of work on adaptive tile rendering - the plan is to increase it to e.g. 67.6 deg to prevent pitch limit based on content insets when padding is small

astojilj commented:
> Note that the failsafe introduced here introduces a change in behavior that’ll be noticeable in the navigation SDK and navigation-related applications. It’s a sound move, but it does put more pressure on us to implement adaptive zoom levels (aka level of detail): #9037.

Thanks a lot for early feedback on this draft PR.
About change in behavior:
I don't think there is a change from API point of view: https://docs.mapbox.com/mapbox-gl-js/api/#cameraoptions describes it "desired pitch value". Didn't see the limit of 60 degrees in documentation.

The change here is done this way to prevent visual change, and allow the largest possible pitch that doesn't show area above horizon.

Agree, there is a pressure to work on #9037 immediately, with or without changes in this PR. I plan to work on #9037 , together with PITCH_MAX constraint removal.

Did I understand the "change in behavior" correctly or there is something additional?

alexshalamov commented on src/mbgl/util/tile_cover.cpp:
```diff
@@ -85,7 +69,7 @@ std::vector<UnwrappedTileID> tileCover(const Point<double>& tl,
-    const int32_t tiles = 1 << z;
+    const int32_t tiles = (1 << z) + 1;
 
     struct ID {
```
We can add constructor and then on line 85
`t.emplace_back(x, y, dx * dx + dy * dy);`

I quickly tested, old version takes 21ns per insertion, with proper emplace 14ns.

astojilj replied:
Thanks. Splitting the tile cover from patch so I'll integrate it there.

astojilj commented:
@1ec5 , @alexshalamov 
PTAL.
As [mentioned](https://github.com/mapbox/mapbox-gl-native/issues/15163#issuecomment-514402526), this is first part of fixing #15195.

@alexshalamov I'll split the tile cover optimization to other patch - here, contributing unit tests for verifying the upcoming patch against current state.

pozdnyakov commented on src/mbgl/map/transform.cpp:
```diff
@@ -576,4 +578,16 @@ LatLng Transform::screenCoordinateToLatLng(const ScreenCoordinate& point, LatLng
+    const auto height = state.size.height;
+    assert(height);
+    // See TransformState::fov description: fov = 2 * arctan((height / 2) / (height * 1.5)).
+    const double fovAboveCenter = std::atan((0.666666 + 0.02) * (0.5 + centerOffsetY / height));
```
could  `0.666666 ` and others be const variables with descriptive names?

astojilj replied:
Fixed in the most recent patch. PTAL.

pozdnyakov commented on src/mbgl/map/transform.cpp:
```diff
@@ -314,6 +312,10 @@ void Transform::flyTo(const CameraOptions &camera, const AnimationOptions &anima
                 util::interpolate(startEdgeInsets.right(), padding.right(), us)
             };
         }
+        auto maxPitch = getMaxPitchForEdgeInsets(state.edgeInsets);
```
could the above statement be calculated only on `state.edgeInsets` change?

astojilj replied:
Use of optionals mandate this implementation: if no insets or no pitch are supplied, means we use current values and still need to set and ensure validity of what is set.

pozdnyakov commented on src/mbgl/map/transform.cpp:
```diff
@@ -576,4 +578,16 @@ LatLng Transform::screenCoordinateToLatLng(const ScreenCoordinate& point, LatLng
+    double centerOffsetY = 0.5 * (insets.top() - insets.bottom()); // See TransformState::getCenterOffset.
+
+    const auto height = state.size.height;
+    assert(height);
```
how is that guarantied?

astojilj replied:
This is invalid state as it would result by division by zero bellow (old and new code).
In production code it couldn't happen:
Map::Impl::Impl calls [Transform::resize](https://github.com/mapbox/mapbox-gl-native/blob/e5431d01817c0546aa0b971a7b0b33c9bbd87077/src/mbgl/map/transform.cpp#L49), where we ensure height != 0.
Constructed value is invalid (0) and could happen with unit tests constructing Transform with no resize call. Fixed some those tests in this patch and assert is added here...

astojilj commented:
@1ec5 ,
Before merging this, I'd like to verify that I've correctly understood and addressed your concerns [here](https://github.com/mapbox/mapbox-gl-native/pull/15195#issuecomment-514400825).

Work on this, and making the max pitch larger, continues in #https://github.com/mapbox/mapbox-gl-native/pull/15230. But, I'd like to push this patch first as it is failsafe, preventing out of memory and block but not altering currently max allowed/possible pitch.

1ec5 commented:
> Work on this, and making the max pitch larger, continues in ##15230. But, I'd like to push this patch first as it is failsafe, preventing out of memory and block but not altering currently max allowed/possible pitch.

This is a good plan, thank you!

astojilj commented:
> Suggested changelog entry:
> 
> * Fixed an issue where it was possible to set the map’s content insets then tilt the map enough to see the horizon, causing performance issues.

Thanks, added.
